### PR TITLE
GUA-748 Write lambda for the activity log

### PR DESCRIPTION
**Pull request by DougsCER**

### CODE REVIEW

File: lambda/write-activity-log/models.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  activities: Activity[];
+  truncated: boolean;
+}
+
+export interface UserActivityLog {
+  TxmaEvent: TxmaEvent;
+  ActivityLogEntry: ActivityLogEntry | undefined;
+}
+
+export const allowedTxmaEvents: Array<string> = [
```
Comment (divyaparameswaran) [inline lambda/write-activity-log/models.ts:36]: Is this needed for the write lambda ?

File: lambda/write-activity-log/write-activity-log.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+          activityLogEntry.user_id  !== undefined &&
+          activityLogEntry.session_id !== undefined &&
+          activityLogEntry.timestamp !== undefined &&
+          activityLogEntry.truncated !== undefined
+      )) {
+      throw new Error(`Could not validate activity log entry ${JSON.stringify(activityLogEntry)}`);
+  }
+};
+
+export const validateActivities = (activities: Activity[]): void => {
```
Comment (divyaparameswaran) [inline lambda/write-activity-log/write-activity-log.ts:33]: Should we call this from `[validateActivityLogEntry](https://github.com/alphagov/di-account-management-backend/pull/104/files#diff-ef509ba4be319ceefbeef69794ebb6e1b02b814eea64b00b7fc3150aa2d08391R22)`

File: lambda/write-activity-log/write-activity-log.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    for (let i = 0; i < activities.length; i += 1) {
```
Comment (divyaparameswaran) [inline lambda/write-activity-log/write-activity-log.ts:34]: Could we use a `for-of` loop here 
Sonar usually complains about these as code smells 
https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&id=di-account-management-backend-write-user-services&open=AYYwXbK98mzb7uqxZ1VG&tab=why

File: lambda/write-activity-log/write-activity-log.ts (hunk 2 above)
Comment (DougsCER) [reply lambda/write-activity-log/write-activity-log.ts:39]: yes

File: lambda/write-activity-log/models.ts (hunk 1 above)
Comment (DougsCER) [reply lambda/write-activity-log/models.ts:36]: i've slimmed the model file down

File: lambda/write-activity-log/write-activity-log.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    }
+}
+
+export const writeActivityLogEntry = async (activityLogEntry: ActivityLogEntry):
+ Promise<PutCommandOutput> => {
+    const { TABLE_NAME } = process.env;
+    const command = new PutCommand({
+      TableName: TABLE_NAME,
+      Item: {
+        user_id: activityLogEntry.user_id,
```
[the remark this answers is not in the document]
Comment (DougsCER) [reply lambda/write-activity-log/write-activity-log.ts]: good suggestion

File: lambda/write-activity-log/write-activity-log.ts (hunk 3 above)
Comment (DougsCER) [reply lambda/write-activity-log/write-activity-log.ts]: good suggestion

File: lambda/write-activity-log/models.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  event_type: string;
+  session_id: string;
+  user_id: string;
+  timestamp: number;
+  activities: Activity[];
+  truncated: boolean;
+}
+
+export interface UserActivityLog {
+  TxmaEvent: TxmaEvent;
```
[the remark this answers is not in the document]
Comment (DougsCER) [reply lambda/write-activity-log/models.ts]: done

File: lambda/write-activity-log/write-activity-log.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+): void => {
+  if (
+    !(
+      activityLogEntry.user_id !== undefined &&
+      activityLogEntry.session_id !== undefined &&
+      activityLogEntry.timestamp !== undefined &&
+      activityLogEntry.truncated !== undefined &&
+      activityLogEntry.event_type !== undefined &&
+      activityLogEntry.activities !== undefined &&
+      activityLogEntry.activities.every(validateActivity)
```
Comment (huwd) [inline lambda/write-activity-log/write-activity-log.ts:38]: nice!

File: .github/workflows/lint-and-test-lambdas.yaml [hunk 7]
```diff
@@ -170,3 +170,27 @@ jobs:
...
+    defaults:
+      run:
+        working-directory: ./lambda/write-activity-log
+    steps:
+      - uses: actions/checkout@8f4b7f84864484a7bf31766abe9204da3cbe65b3 # pin@v3
+
+      - name: Install Node.js
+        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # pin@v3
+        with:
+          node-version-file: ".nvmrc18"
```
Comment (huwd) [inline .github/workflows/lint-and-test-lambdas.yaml:186]: Note to self, we should really get the rest of these ones on node 18 and then turn these new ones back to using .nvmrc

**Review by huwd — COMMENTED** [review]
I think it looks great, I'd squash these two commits down to one before merging, but ✅ happy to go after that?

**Review by huwd — APPROVED** [review]

File: .github/workflows/sam-app-post-merge-actions.yaml [hunk 8]
```diff
@@ -56,6 +56,11 @@ jobs:
       - name: Test query-activity-log
         run: npm ci && npm test
         working-directory: ./lambda/query-activity-log
+      
+      - name: Test write-activity-log
+        run: npm ci && npm test
+        working-directory: ./lambda/write-activity-log
+
```
Comment (divyaparameswaran) [inline .github/workflows/sam-app-post-merge-actions.yaml:63]: Could we please include  write-activity-log to the sonar workflows as well

Comment (huwd) [reply .github/workflows/sam-app-post-merge-actions.yaml:63]: Oh that's a good catch, I'd be happy for this to go in here, or to make this a follow up task.
I know merging this unblocks you on delivery a bit right Dougal?

And there's nothing much we lose from adding this say... later this week over ensuring it's there at the start right @divyaparameswaran?

So long as we don't forget I'd be happy. I'm still getting relatively variable mileage out of using sonar 🤔

**Pull request merged**